### PR TITLE
Update the `Testing and Specifying the API` page

### PR DESCRIPTION
**Pull request by meyerbaptiste**

### CODE REVIEW

File: distribution/testing.md [hunk 1]
```diff
@@ -1,194 +1,323 @@
...
+
+namespace App\Tests;
+
+use Hautelook\AliceBundle\PhpUnit\ReloadDatabaseTrait;
+use Symfony\Bundle\FrameworkBundle\Client;
+use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
+
+class ApiTest extends WebTestCase
+{
+    use ReloadDatabaseTrait;
```
Comment (meyerbaptiste) [inline distribution/testing.md]: n.b. We need to find a proper fix for traits work with `@depends` without the `setUpBeforeClass` hack and we need help from @theofidry for a new release of `hautelook/alice-bundle` which integrates https://github.com/hautelook/AliceBundle/pull/441.

File: distribution/testing.md [hunk 2]
```diff
@@ -1,194 +1,323 @@
...
-  I need to be able to retrieve, create, update and delete them through the API.
-
-  # the "@createSchema" annotation provided by API Platform creates a temporary SQLite database for testing the API
-  @createSchema
-  Scenario: Create a book
-    When I add "Content-Type" header equal to "application/ld+json"
-    And I add "Accept" header equal to "application/ld+json"
-    And I send a "POST" request to "/books" with body:
-    """
+Now that you have a functional API, it might be interesting to write some tests to ensure your API have no potentially
```
Comment (alanpoulain) [inline distribution/testing.md]:
```suggestion
Now that you have a functional API, it might be interesting to write some tests to ensure your API have no potential
```

File: distribution/testing.md [hunk 3]
```diff
@@ -1,194 +1,323 @@
...
+You can now try to load your fixtures in your database with the following command:
+
+    $ docker-compose exec php bin/console hautelook:fixtures:load --env=test
+
+To learn more about fixtures, take a look at the documentation of [Alice](https://github.com/nelmio/alice/blob/master/README.md#table-of-contents)
+and [AliceBundle](https://github.com/hautelook/AliceBundle/blob/master/README.md).
+
+## Writing Functional Tests
+
+Now you have some data fixtures for your API, you are ready to write functional tests with [PHPUnit](https://phpunit.de/index.html).
```
Comment (alanpoulain) [inline distribution/testing.md]:
```suggestion
Now that you have some data fixtures for your API, you are ready to write functional tests with [PHPUnit](https://phpunit.de/index.html).
```

File: distribution/testing.md [hunk 4]
```diff
@@ -1,194 +1,323 @@
...
+}
+```
 
-    $ docker-compose exec php vendor/bin/behat
+As you can see, the example uses the [trait `ReloadDatabaseTrait`](https://github.com/hautelook/AliceBundle#database-testing)
+from [AliceBundle](https://github.com/hautelook/AliceBundle/blob/master/README.md) which will purge the database and load
+fixtures before every tests. Because of this, you can run your tests without worrying about fixtures.
 
-Everything should be green now. Your Linked Data API is now specified and tested thanks to Behat!
+All you have to do now is run your tests:
```
Comment (alanpoulain) [inline distribution/testing.md]:
```suggestion
All you have to do now is to run your tests:
```

File: distribution/index.md [hunk 5]
```diff
@@ -331,7 +331,7 @@ class Review
     /**
      * @var \DateTimeInterface The date of publication of this review.
      *
-     * @ORM\Column(type="datetime_immutable")
+     * @ORM\Column(type="datetime")
```
Comment (dunglas) [inline distribution/index.md]: Why this change? Using the immutable version is preferred.

File: distribution/testing.md [hunk 6]
```diff
@@ -1,194 +1,241 @@
...
+App\Entity\Review:
+    review_{1..20}:
+        rating: <numberBetween(0, 5)>
+        body: <text()>
+        author: <name()>
+        publicationDate: <dateTime()>
+        book: '@book_<numberBetween(1, 10)>'
+```
+
+You can now try to load your fixtures in your database with the following command:
```
Comment (dunglas) [inline distribution/testing.md]:
```suggestion
You can now load your fixtures in the database with the following command:
```

File: distribution/testing.md [hunk 7]
```diff
@@ -1,194 +1,241 @@
...
+
+    $ docker-compose exec php bin/console hautelook:fixtures:load --env=test
```
Comment (dunglas) [inline distribution/testing.md]:
```suggestion
    $ docker-compose exec php bin/console hautelook:fixtures:load
```

File: distribution/testing.md [hunk 8]
```diff
@@ -1,194 +1,241 @@
...
+
+## Writing Functional Tests
+
+Now that you have some data fixtures for your API, you are ready to write functional tests with [PHPUnit](https://phpunit.de/index.html).
+
+Install the Symfony test pack which includes [PHPUnit Bridge](https://symfony.com/doc/current/components/phpunit_bridge.html):
+
+    $ docker-compose exec php composer require --dev test-pack
+
+Your API is ready to be functionally tested. Create your test classes under the `tests` directory.
```
Comment (dunglas) [inline distribution/testing.md]:
```suggestion
Your API is ready to be functionally tested. Create your test classes under the `tests/` directory.
```

File: distribution/testing.md [hunk 9]
```diff
@@ -1,194 +1,241 @@
...
+
+Here is an example of functional tests specifying the behavior of [the bookstore API you created in the tutorial](index.md):
+
+```php
+<?php
+
+namespace App\Tests;
```
Comment (dunglas) [inline distribution/testing.md]:
```suggestion
// api/tests/ApiTest.php
namespace App\Tests;
```

File: distribution/testing.md [hunk 10]
```diff
@@ -1,194 +1,241 @@
...
+{
+    use RefreshDatabaseTrait;
+
+    /** @var Client */
+    protected $client;
+
+    /**
+     * Retrieves the documentation.
+     */
+    public function testRetrieveTheDocumentation(): void
```
Comment (dunglas) [inline distribution/testing.md]: I would move this test at the end, testing the HTML docs is a edge case.

File: distribution/testing.md [hunk 11]
```diff
@@ -1,194 +1,241 @@
...
+        return $this->client->getResponse();
     }
-    """
+}
 ```
 
-The API Platform flavor of Behat also comes with a temporary SQLite database dedicated to tests. It works out of the box.
-
-Clear the cache of the `test` environment:
+As you can see, the example uses the [trait `RefreshDatabaseTrait`](https://github.com/hautelook/AliceBundle#database-testing)
```
Comment (dunglas) [inline distribution/testing.md]: Can you explain quickly that is uses transactions.

File: distribution/testing.md [hunk 12]
```diff
@@ -1,194 +1,241 @@
...
-* [PHP Matcher](https://github.com/coduo/php-matcher): the Swiss Army knife of JSON document testing.
-
-## Running Unit Tests with PHPUnit
+* [ApiTestCase](https://github.com/lchrusciel/ApiTestCase), a handy [PHPUnit](https://phpunit.de/index.html) test case
+  for going further by testing JSON and XML APIs in your Symfony applications;
+* [Behat](http://behat.org/en/latest/) and its [Behatch extension](https://github.com/Behatch/contexts), a
+  [Behavior-Driven development](https://en.wikipedia.org/wiki/Behavior-driven_development) framework to write the API
+  specification as user stories and in natural language then execute these scenarios against the application to validate
+  its behavior;
+* [Blackfire Player](https://blackfire.io/player), a nice DSL to crawl HTTP services, assert responses, and extract data
```
Comment (dunglas) [inline distribution/testing.md]: Could you add a link to the example usage in the demo repo please?

File: distribution/testing.md [hunk 13]
```diff
@@ -1,194 +1,241 @@
...
 
-Then run:
+    $ docker-compose exec php bin/phpunit
 
-    $ docker-compose exec php vendor/bin/behat
+If everything is working properly, you should see `OK (4 tests, 21 assertions)`. Your Linked Data API is now specified
+and tested thanks to [PHPUnit](https://phpunit.de/index.html)!
 
-Everything should be green now. Your Linked Data API is now specified and tested thanks to Behat!
+### Finding Alternatives
```
Comment (dunglas) [inline distribution/testing.md]:
```suggestion
### Additional and Alternative Testing Tools
```

File: distribution/testing.md [hunk 14]
```diff
@@ -1,194 +1,241 @@
...
 To install [PHPUnit](https://phpunit.de/) test suite, execute the following command:
 
-    $ docker-compose exec php composer require --dev symfony/phpunit-bridge
+    $ docker-compose exec php composer require --dev test-pack
 
 To run your [PHPUnit](https://phpunit.de/) test suite, execute the following command:
 
     $ docker-compose exec php bin/phpunit
+
+Take a look at [the Symfony documentation about testing](https://symfony.com/doc/current/testing.html) to learn how to
```
Comment (dunglas) [inline distribution/testing.md]: I would keep only this part, as we already explain how to install PHPUnit in the previous section.

File: distribution/index.md (hunk 5 above)
Comment (meyerbaptiste) [reply distribution/index.md]: There is no formatter for the immutable version in Faker.

**Comment by meyerbaptiste** [comment]
Done @dunglas.

**Comment by bpolaszek** [comment]
Just FYI the JWT section of the docs still exposes a Behat example: https://github.com/api-platform/docs/blob/b739abfaec96f8d921da3a86fc1ae19eb64fa73d/core/jwt.md

Thank you!  🙂
Ben

**Comment by dunglas** [comment]
Thanks @meyerbaptiste
